ISPN-7986 Hot RodTransaction (Server only)

pruivo opened the pull request:
https://issues.jboss.org/browse/ISPN-7986

Hot Rod server is able to handle transaction from a client.
The Hot Rod client will be implemented later.
Optimistic and Total Order transactions aren't supported yet due to a bug between hot rod operations and versioning.

question: should the protocol be increase to 2.7? or only when the client is implemented?

wburns commented:
I would personally say we may want to bump the protocol to 3.0. Adding transactions is a pretty big change to me. Anyone else have an opinion?

This seems like it would be 9.2.0 as well right?

pruivo commented:
@wburns I would like to have this PR in 9.1. Although the server can handle TX, the client can't so it is harmless. The version could be incremented when the client is implemented as well.

gustavonalle commented:
IMO the version should be increased now to 2.7, even without the clients. Great opportunity to test older clients (2.6) with newer servers (2.7)  :)

wburns commented:
I am guessing we will need a CR2 build then to contain this. This seems like an awful lot of changes for a Final build.

wburns commented:
Although @tristantarrant would control that.

tristantarrant commented:
Too late for 9.1. this will land early on in 9.2

On 27 Jun 2017 3:59 pm, "William Burns" <notifications@github.com> wrote:

> Although @tristantarrant <https://github.com/tristantarrant> would
> control that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/infinispan/infinispan/pull/5240#issuecomment-311367008>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAX9Huy1tg7CfhPVMv-iuU-Rf2Yd-FJvks5sIQrYgaJpZM4OGpQ2>
> .
>

rvansa commented:
Wrt versions - I second 2.7 (server side). With the advent of builiding blocks (counters, multimaps...) I would like to see 3.0 being modular to accommodate these.

tristantarrant commented:
I've never liked the single global protocol version approach. It would make
more sense to leave the global version just as a way to indicate the
general operation framing and format, version the individual ops (á la
JGroups) and return the supported (versioned) ops as part of PING.
e.g. PUT PUT2 GET ...

On 4 Jul 2017 08:30, "Radim Vansa" <notifications@github.com> wrote:

> Wrt versions - I second 2.7 (server side). With the advent of builiding
> blocks (counters, multimaps...) I would like to see 3.0 being modular to
> accommodate these.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/infinispan/infinispan/pull/5240#issuecomment-312791862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAX9HhtFRZxRQua74bhoq01xCTo15g2rks5sKdvzgaJpZM4OGpQ2>
> .
>

rvansa commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/tx/TxState.java:
```diff
@@ -0,0 +1,183 @@
+   private static List<WriteCommand> unmarshallModifications(ObjectInput input)
+         throws IOException, ClassNotFoundException {
+      List<WriteCommand> modifications = MarshallUtil.unmarshallCollection(input, ArrayList::new);
+      return modifications == null ? null : Collections.unmodifiableList(modifications);
```
do we need to be *that* defensive when the `getModifications` is package-private?

pruivo replied:
ok, no more defensive code :)

rvansa commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/tx/TxState.java:
```diff
@@ -0,0 +1,183 @@
+            null;
+   }
+
+   public enum TxStatus {
```
I don't fancy another enum describing tx state... can't you use `javax.transaction.Status` codes and a little switch in static `canAdvance`?

pruivo replied:
ok

rvansa commented on documentation/src/main/asciidoc/user_guide/server_protocols_hotrod_wire.adoc:
```diff
@@ -2013,6 +2014,151 @@ Here are the current values for each flag: +
+|vInt
+|The number of keys
+
+3+^|For each key
```
You should mention that the server requires all keys to be distinct here.

rvansa commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/tx/TransactionDecodeContext.java:
```diff
@@ -0,0 +1,105 @@
+      this.transactionTable = registry.getComponent(TransactionTable.class);
+      this.clusteringDependentLogic = registry.getComponent(ClusteringDependentLogic.class);
+      this.serverTransactionTable = registry.getComponent(ServerTransactionTable.class);
+      this.tm = (EmbeddedTransactionManager) cache.getTransactionManager();
```
I don't really understand *why* we need a JTA transaction manager at all, when we force it to be the `EmbeddedTransactionManager` anyway. I think the transaction manager should be only on the user-code side, not in server. Please elaborate.

pruivo replied:
All the core code interacts with TransactionManager and Transaction interface. I don't see any other way around exception re-write the core... What idea do you have in mind?

rvansa replied:
What core code? Interceptors should interact only with `TxInvocationContext`, only CacheWriterInterceptor and CacheNotifierImpl suspend the running tx. I've found that applying state in `StateConsumerImpl` creates it's own transaction using the TM, but I am not sure how much that'd be necessary there.

pruivo replied:
but the CacheImpl interacts with the TransactionManager to map the Transaction > LocalTransaction.

rvansa replied:
You don't have to use CacheImpl directly in the server; I suggest overriding `DecoratedCache.readContext`/`writeContext` and pass the transaction identifier to the created context directly.

pruivo replied:
it isn't possible. I can't decorate the cache and I really don't want to re-implement the cache.
```
java.lang.ClassCastException: org.infinispan.cache.impl.EncoderCache cannot be cast to org.infinispan.cache.impl.CacheImpl
        at org.infinispan.cache.impl.DecoratedCache.<init>(DecoratedCache.java:70) ~[infinispan-core-9.1.1-SNAPSHOT.jar:9.1.1-SNAPSHOT]
        at org.infinispan.cache.impl.DecoratedCache.<init>(DecoratedCache.java:56) ~[infinispan-core-9.1.1-SNAPSHOT.jar:9.1.1-SNAPSHOT]
```

pruivo commented on documentation/src/main/asciidoc/user_guide/server_protocols_hotrod_wire.adoc:
```diff
@@ -129,15 +129,16 @@ When topology-aware or hash-distribution-aware clients will send 0 until they
-| Transaction Id      | byte array |  The byte array uniquely identifying the
-transaction associated to this call. Its length is determined by the
-transaction type. If transaction type is 0, no transaction id will be present.
+//TODO this isn't implemented. Implement it will break compatbility with older clients
```
there isn't nothing to remove. Decoder10 reads the transactionType but in Decoder2x it don't read it. 
I checked the client code as well and only the Codec13 sends the transactionType. Starting from Codec20, it isn't sent.

pruivo commented on documentation/src/main/asciidoc/user_guide/server_protocols_hotrod_wire.adoc:
```diff
@@ -2013,6 +2014,151 @@ Here are the current values for each flag: +
+|Byte array containing the key
+
+|Version Read
+|long
```
what? why complicated? The `RemoteCache` interface uses long (e.g. `removeWithVersion()`), and I don't think we are going to change it. The only issue we have is that the hot rod version is overwriting the entry version (optimistic transactions) and it shouldn't.

pruivo commented on documentation/src/main/asciidoc/user_guide/server_protocols_hotrod_wire.adoc:
```diff
@@ -2013,6 +2014,151 @@ Here are the current values for each flag: +
+
+|Lifespan
+|vInt
+|Number of seconds that a entry during which the entry is allowed to life.
```
ok

pruivo commented on documentation/src/main/asciidoc/user_guide/server_protocols_hotrod_wire.adoc:
```diff
@@ -2013,6 +2014,151 @@ Here are the current values for each flag: +
+
+.Response from prepare, commit and rollback request.
+
+Response 0x3C, 0x3E and 0x40
```
I think so. In the docs there are a bunch of operations with the same response format and different response code (e.g. putIfAbsent, replace)

pruivo commented on documentation/src/main/asciidoc/user_guide/server_protocols_hotrod_wire.adoc:
```diff
@@ -2013,6 +2014,151 @@ Here are the current values for each flag: +
+
+.XID Formtat
+
+The XID in the requests has the following format:
```
why? the transaction doesn't live in the server. only in the client. The script wouldn't have access to the transaction context (keys read/written by the tx) in the server.

rvansa replied:
And that's part of the problem. But running a script within the scope of a transaction and not participating in that transaction is not correct. If you don't want to deal with that, clearly describe the limitations in docs.

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -58,6 +75,168 @@ public RequestParameters getParams() {
+   /**
+    * Handles a prepare request from a client
+    */
+   Response prepareTransaction() throws Exception {
```
ok

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -58,6 +75,168 @@ public RequestParameters getParams() {
+
+      PrepareTransactionContext context = (PrepareTransactionContext) operationDecodeContext;
+      if (context.isEmpty()) {
+         //the client can optimize and avoid contacting the server when no data is written.
```
Our implementation would sent it but I'm not preventing a user implementation to send an empty modification set.

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -58,6 +75,168 @@ public RequestParameters getParams() {
+      try {
+         txContext.perform();
+      } catch (HeuristicMixedException e) {
+         return createTransactionResponse(header, XAException.XA_HEURMIX);
```
I'll double check but I'm pretty sure that the core logs the exception.

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -58,6 +75,168 @@ public RequestParameters getParams() {
+      //forces the write-lock. used by pessimistic transaction. it ensures the key is not written after is it read and validated
+      //optimistic transaction will use the write-skew check.
+      AdvancedCache<byte[], byte[]> readCache = cache.withFlags(Flag.FORCE_WRITE_LOCK);
+      boolean failed = false;
```
ok

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -301,6 +484,36 @@ ComponentRegistry getCacheRegistry(String cacheName) {
+         if (isTrace) {
+            log.tracef("Key didn't exist for operation %s. Entry is %s", write, entry);
+         }
+         return entry == null;
```
ok

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -301,6 +484,36 @@ ComponentRegistry getCacheRegistry(String cacheName) {
+      if (isTrace) {
+         log.tracef("Checking version for operation %s. Entry is %s", write, entry);
+      }
+      return entry != null && write.versionRead == extractVersion(entry.getMetadata().version());
```
I'll use the extractVersion until the issue is fixed. Otherwise, it won't be compliant since the getWithMetadata uses this method to send the version back to the client (and it is this version that is sent to validate)

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/Decoder2x.java:
```diff
@@ -332,32 +345,120 @@ public void customReadHeader(HotRodHeader header, ByteBuf buffer, CacheDecodeCon
+   }
+
+   private Optional<TransactionWrite> readTransactionWrite(HotRodHeader header, ByteBuf byteBuf) {
+      return readMaybeRangedBytes(byteBuf).flatMap(key ->
```
I don't like it either but it seems the way to avoid blocking the reader thread for data.

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/Decoder2x.java:
```diff
@@ -332,32 +345,120 @@ public void customReadHeader(HotRodHeader header, ByteBuf buffer, CacheDecodeCon
+                     if (ControlByte.REMOVE_OP.hasFlag(control)) {
+                        return Optional.of(new TransactionWrite(key, versionRead, control, null, null, null));
+                     }
+                     return readLifespanOrMaxIdle(byteBuf, hasFlag(header, ProtocolFlag.DefaultLifespan))
```
yes, I'm an idiot :)

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/Decoder2x.java:
```diff
@@ -332,32 +345,120 @@ public void customReadHeader(HotRodHeader header, ByteBuf buffer, CacheDecodeCon
+   }
+
+   private Optional<Integer> readLifespanOrMaxIdle(ByteBuf buffer, boolean useDefault) {
+      return readMaybeVInt(buffer).map(stream -> {
```
copy-pasted... I'll change it.

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/tx/ServerTransactionOriginatorChecker.java:
```diff
@@ -0,0 +1,50 @@
+   }
+
+   private boolean isNonClientTransaction(GlobalTransaction gtx) {
+      return globalTxTable.values().stream()
```
a client address where?

rvansa replied:
`GlobalTransaction.getAddress()` would return an instance of `Address` that does not match to any node address.

pruivo replied:
that will break the transactions. GlobalTransaction needs to be unique and your suggestion would make all nodes to create the GlobalTransaction with the same Address. 
I guess I can create a client address per server/cache manager if that is ok for you. I'm open to other suggestions as well :)

rvansa replied:
Yep, that sounds good.

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/tx/TxFunctions.java:
```diff
@@ -0,0 +1,114 @@
+ * @author Pedro Ruivo
+ * @since 9.1
+ */
+public class TxFunctions {
```
I can try but I found out it sends the function when topology is changing...

rvansa replied:
Ah, when the cache has joined the replicated cache but did not receive all data yet, it probably has to execute a stream operation remotely. In any way, doing the transaction originator check using a RPC is a no-go IMO.

pruivo commented:
Updated:
* advanced version to 2.7
* refactor the expiration decoding  (it was bugged)

rvansa commented:
Re [entry version as long]: In general, Infinispan supports and `EntryVersion` implementations; you're right that with current `RemoteCache` the full version info is omitted. It's not 'only overwriting the entry version', it's breaking any use of versions right now.
As I've said, we need a plan for `EntryVersion` in general. Either we enforce `NumericVersion` everywhere (which would be unfortunate IMO, as we might find good use cases for that - e.g. SC uses SimpleClusteredVersion, too - I know we currently don't support that together with Hot Rod, but...) or we accept other version types at least in the protocol, eventually adjusting RemoteCache API as well (temporarily scrambling other versions to long, but not ignoring parts). Do you see any other options?

pruivo commented:
the plan is to remove it from the public interface. We support EntryVersion but we don't. If you try to configure an optimistic tx cache with a custom VersionGenerator, it will give you an error. If you configure a custom VersionGenerator with any other type of cache, it will be simply ignored. It shouldn't be public anyway. 
And the eventual consistency stuff will probably needs another implementation....

pruivo commented:
Updated:
* new commit with GlobalTransaction created with a ClientAddress. It avoid iterate over the values to find if it is a client transaction or not.
* new commit with removal of TransactionManager... it is a bit hacky.

pruivo commented:
updated:
* new commit: added factory method to tests (test with different transaction modes)

ps. commits will be squashed when review is finished.

tristantarrant commented:
Temporarily closing because of CI overload

pruivo commented:
reopened PR for 9.2 alpha

pruivo commented:
rebased

galderz commented:
New failure?

```
org.infinispan.server.test.jmx.management.JmxManagementIT.testCacheManagerAttributes
```

pruivo commented:
should be fixed now :+1:

galderz commented:
Integrating, thanks @pruivo! :)
